beets/dbcore/db.py

```diff
@@ -47,6 +49,15 @@
 if TYPE_CHECKING:
     from types import TracebackType
 
+    from .query import SQLiteType
+
+    D = TypeVar("D", bound="Database", default=Any)
+else:
+    D = TypeVar("D", bound="Database")
```

Comment:
Do we need to set `default` here? (I wouldn't have expected it: `Model` is abstract, so we can't ever create `Model` objects, where `D` would not be specified.)

Wouldn't it also be easier to do this unconditionally? (If `not TYPE_CHECKING`, `D` isn't really used anyway.)

```suggestion
D = TypeVar("D", bound="Database")
```

or, if we do need the default, make it more specific?

```suggestion
D = TypeVar("D", bound="Database", default="Database")
```

Reply:
We use plain `Model` in type definitions without any type on it, so I'd expect mypy to complain. Have just removed the default and ran the analysis:

```
beets/dbcore/query.py:94: error: Missing type parameters for generic type "Model"  [type-arg]
beets/dbcore/query.py:159: error: Missing type parameters for generic type "Model"  [type-arg]
beets/dbcore/query.py:199: error: Missing type parameters for generic type "Model"  [type-arg]
beets/dbcore/query.py:394: error: Missing type parameters for generic type "Model"  [type-arg]
beets/dbcore/query.py:533: error: Missing type parameters for generic type "Model"  [type-arg]
beets/dbcore/query.py:572: error: Missing type parameters for generic type "Model"  [type-arg]
beets/dbcore/query.py:582: error: Missing type parameters for generic type "Model"  [type-arg]
beets/dbcore/query.py:608: error: Missing type parameters for generic type "Model"  [type-arg]
beets/dbcore/query.py:627: error: Missing type parameters for generic type "Model"  [type-arg]
beets/dbcore/query.py:637: error: Missing type parameters for generic type "Model"  [type-arg]
beets/dbcore/query.py:828: error: Missing type parameters for generic type "Model"  [type-arg]
beets/dbcore/query.py:1001: error: Missing type parameters for generic type "Model"  [type-arg]
beets/dbcore/db.py:89: error: Missing type parameters for generic type "Model"  [type-arg]
beets/dbcore/db.py:125: error: Missing type parameters for generic type "Model"  [type-arg]
beets/dbcore/db.py:156: error: Missing type parameters for generic type "Model"  [type-arg]
beets/dbcore/db.py:346: error: Missing type parameters for generic type "Model"  [type-arg]
beets/dbcore/db.py:419: error: Missing type parameters for generic type "Model"  [type-arg]
beets/dbcore/db.py:735: error: Missing type parameters for generic type "Model"  [type-arg]
beets/dbcore/db.py:752: error: Missing type parameters for generic type "Model"  [type-arg]
beets/dbcore/db.py:1012: error: Missing type parameters for generic type "Model"  [type-arg]
beets/dbcore/queryparse.py:115: error: Missing type parameters for generic type "Model"  [type-arg]
beets/dbcore/queryparse.py:183: error: Missing type parameters for generic type "Model"  [type-arg]
beets/dbcore/queryparse.py:200: error: Missing type parameters for generic type "Model"  [type-arg]
beets/dbcore/queryparse.py:231: error: Missing type parameters for generic type "Model"  [type-arg]
beets/dbcore/queryparse.py:250: error: Missing type parameters for generic type "Model"  [type-arg]
```

Reply:
And that's what we see if we try to set the default to `Database`:

(see my message above for my theory regarding this 😅)

```
beets/dbcore/db.py:367: error: Argument 1 to "LazyConvertDict" has incompatible type "Model[D]"; expected "Model[Database]"  [arg-type]
beets/dbcore/db.py:368: error: Argument 1 to "LazyConvertDict" has incompatible type "Model[D]"; expected "Model[Database]"  [arg-type]
beets/dbcore/db.py:432: error: Incompatible return value type (got "Model[D]", expected "Model[Database]")  [return-value]
beets/dbcore/db.py:432: note: Perhaps you need a type annotation for "new"? Suggestion: "Model[Database]"
beets/dbcore/db.py:449: error: Invalid self argument "type[Model[D]]" to class attribute function "_getters" with type "Callable[[type[Model[Database]]], Any]"  [misc]
beets/dbcore/db.py:517: error: Invalid self argument "type[Model[D]]" to class attribute function "_getters" with type "Callable[[type[Model[Database]]], Any]"  [misc]
beets/dbcore/db.py:686: error: Argument 1 has incompatible type "Model[D]"; expected "Model[Database]"  [arg-type]
Found 6 errors in 1 file (checked 1 source file)
```

Reply:
Ok, fair enough: I've also been playing around with this locally for a bit. My impression is that in some cases (e.g. `dbcore.db.FormattedMapping`), we really want `Model[Any]` since the resulting type really behaves the same for any `Model` subclass. In other cases (e.g. some `Database` methods), we could be more precise about the specific subclass (because there, we know that `D = Self`). Yet in other cases, we could get stricter typing by making more classes generic (e.e. `dbcore.db.Results`), but that would make things more complicated, and it also seems that I run into https://github.com/python/typing/issues/548.

I'd conclude to keep the `default=Any`, but try to remember that there are places where we could revisit and tighten typing at some point.